pkg/submitter/submitter.go

```diff
@@ -0,0 +1,573 @@
+						"err", err,
+					)
+					if len(val) == 0 {
+						level.Error(s.logger).Log("msg", "0 value being submitted")
+						s.currentValues[i] = big.NewInt(0)
+					}
+					continue
+				}
+				return nil, errors.Errorf("no value in database,  reg id:%v", challenge.RequestIDs[i].Uint64())
+			}
+		}
+		s.currentValues[i] = value
+	}
+
+	// Submit the solution.
+	tx, err := rpc.SubmitContractTxn(ctx, s.logger, s.cfg, s.proxy, s.client, s.contractInstance, s.account, "submitSolution", s.submit)
+	if err == nil {
+		return tx, nil
+	}
+	return nil, err
+}
+
+func (s *Submitter) getMinerStatus() int64 {
+	// Check if the staked account is in dispute before sending a transaction.
+	statusID, _, err := s.contractInstance.GetStakerInfo(&bind.CallOpts{}, s.account.Address)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "getting staker info from contract", "pubkey", s.account.Address.String())
+		return -1
+	}
+	return statusID.Int64()
+}
+
+func (s *Submitter) submit(ctx context.Context, contract tellorCommon.ContractInterface) (*types.Transaction, error) {
+
+	txn, err := contract.SubmitSolution(
+		s.currentNonce,
+		s.currentChallenge.RequestIDs,
+		s.currentValues)
+	if err != nil {
+		return nil, err
+	}
+
+	return txn, err
+}
+
+func (s *Submitter) lastSubmit() (time.Duration, *time.Time, error) {
+	address := "000000000000000000000000" + s.account.Address.Hex()[2:]
+	decoded, err := hex.DecodeString(address)
+	if err != nil {
+		return 0, nil, errors.Wrapf(err, "decoding address")
+	}
+	last, err := s.contractInstance.GetUintVar(nil, rpc.Keccak256(decoded))
+
+	if err != nil {
+		return 0, nil, errors.Wrapf(err, "getting last submit time for:%v", s.account.Address.String())
+	}
+	// The Miner has never submitted so put a timestamp at the beginning of unix time.
+	if last.Int64() == 0 {
+		last.Set(big.NewInt(1))
+	}
+
+	lastInt := last.Int64()
+	now := time.Now()
+	var lastSubmit time.Duration
+	var tm time.Time
+	if lastInt > 0 {
+		tm = time.Unix(lastInt, 0)
+		lastSubmit = now.Sub(tm)
+	}
+
+	return lastSubmit, &tm, nil
+}
+
+// saveGasUsed calculates the price for a given slot.
+// Since the transaction doesn't include the slot number it gets the slot number
+// as soon as the transaction passes and
+// saves it in the database for profit calculations.
+// TODO[Krasi] To be more detirministic and simplify this
+// should get the `_SLOT_PROGRESS` and `gasUsed` from the `NonceSubmitted` event.
+// At the moment there is a slight chance of a race condition if
+// another transaction has passed between checking the transaction cost and
+// checking the `_SLOT_PROGRESS`
+// Tracking issue https://github.com/tellor-io/TellorCore/issues/101
+func (s *Submitter) saveGasUsed(ctx context.Context, tx *types.Transaction) {
+	go func(tx *types.Transaction) {
+		receipt, err := bind.WaitMined(ctx, s.client, tx)
+		if err != nil {
+			level.Error(s.logger).Log("msg", "transaction result for calculating transaction cost", "err", err)
```

Comment:
should return here otherwise it panics
```suggestion
			level.Error(s.logger).Log("msg", "transaction result for calculating transaction cost", "err", err)
			return
```